### PR TITLE
romToRam cache follow up

### PR DESCRIPTION
**Pull request by theresa-m**
- Declare variables first in romToRam entry removal
- Correctly free romToRam hash table and mutex

https://github.com/eclipse-openj9/openj9/pull/20169#discussion_r1766925674

Port to 0.48: https://github.com/eclipse-openj9/openj9/pull/20197

### CODE REVIEW

File: runtime/util/hshelp.c [hunk 1]
```diff
@@ -2643,8 +2643,8 @@ recreateRAMClasses(J9VMThread * currentThread, J9HashTable * classHashTable, J9H
 		if (!fastHCR) {
 			vmFuncs->hashClassTableDelete(classLoader, J9UTF8_DATA(className), J9UTF8_LENGTH(className));
 			if ((NULL != vm->sharedClassConfig) && (NULL != vm->sharedClassConfig->romToRamHashTable)) {
-				omrthread_rwmutex_enter_write(vm->sharedClassConfig->romToRamHashTableMutex);
 				RomToRamEntry entry;
+				omrthread_rwmutex_enter_write(vm->sharedClassConfig->romToRamHashTableMutex);
 				entry.ramClass = originalRAMClass;
```
Comment (keithc-ca) [inline runtime/util/hshelp.c:2648]: Suggest swapping these two lines (initializing `entry` before acquiring the mutex).

**Review by hangshao0 — APPROVED** [review]

File: runtime/shared_common/shrinit.cpp [hunk 2]
```diff
@@ -4329,15 +4327,13 @@ j9shr_shutdown(J9JavaVM *vm)
 		if (utfHashTable) {
 			hashTableFree(utfHashTable);
 		}
-		if (NULL != romToRamHashTable) {
-			omrthread_rwmutex_enter_write(romToRamHashTableMutex);
+		if (NULL != config->romToRamHashTable) {
```
[the remark this answers is not in the document]
Comment (keithc-ca) [reply runtime/shared_common/shrinit.cpp:4332]: Suggest this should use `compareAndSwapUDATA()` to ensure that exactly one thread handles releasing the `romToRamHashTable` and related resources.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/shared_common/shrinit.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/shared_common/shrinit.cpp:4332]: `config` is already freed before this at line 4307.

**Review by hangshao0 — APPROVED** [review]

**Comment by theresa-m** [comment]
I have confirmed that this fixes the previous failures http://vmfarm.rtp.raleigh.ibm.com/build_info.php?build_id=78298. Any other comments @keithc-ca ?

File: runtime/shared_common/shrinit.cpp [hunk 3]
```diff
@@ -4303,6 +4303,14 @@ j9shr_shutdown(J9JavaVM *vm)
 		if (config->jclCacheMutex) {
 			omrthread_monitor_destroy(config->jclCacheMutex);
 		}
+
+		if (compareAndSwapUDATA((uintptr_t *)&config->romToRamHashTable, (uintptr_t)romToRamHashTable, (uintptr_t)NULL)) {
+			config->romToRamHashTableMutex = NULL;
+			omrthread_rwmutex_enter_write(romToRamHashTableMutex);
```
Comment (keithc-ca) [inline runtime/shared_common/shrinit.cpp:4309]: The local variables `romToRamHashTable` and `romToRamHashTableMutex` should be declared and initialized here; the return value should be compared explicitly with `NULL` and the long line folded:
```c
		J9HashTable *romToRamHashTable = config->romToRamHashTable;
		if (NULL != compareAndSwapUDATA(
				(uintptr_t *)&config->romToRamHashTable,
				(uintptr_t)romToRamHashTable,
				(uintptr_t)NULL)
		) {
			omrthread_rwmutex_t romToRamHashTableMutex = config->romToRamHashTableMutex;
			config->romToRamHashTableMutex = NULL;
			omrthread_rwmutex_enter_write(romToRamHashTableMutex);
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/shared_common/shrinit.cpp (hunk 3 above)
Comment (theresa-m) [reply runtime/shared_common/shrinit.cpp:4309]: Shouldn't `J9HashTable *romToRamHashTable` be declared at the beginning of the block where it is now?

File: runtime/shared_common/shrinit.cpp [hunk 4]
```diff
@@ -4303,6 +4303,17 @@ j9shr_shutdown(J9JavaVM *vm)
 		if (config->jclCacheMutex) {
 			omrthread_monitor_destroy(config->jclCacheMutex);
 		}
+
+		if (0 != compareAndSwapUDATA((uintptr_t *)&config->romToRamHashTable,
+			(uintptr_t)romToRamHashTable,
+			(uintptr_t)0)
+		) {
```
Comment (keithc-ca) [inline runtime/shared_common/shrinit.cpp:4310]: Please move the locals here as requested (it narrow the gap between the read and test of `romToRamHashTable` and there's no need to read `romToRamHashTableMutex` unless we're going to use and clear it).
When a function call is split, each argument goes on a line by itself.

Comment (keithc-ca) [reply runtime/shared_common/shrinit.cpp:4309]: > When a function call is split, each argument goes on a line by itself.

See https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#function-parameter-list-does-not-fit-on-one-line.

Comment (theresa-m) [reply runtime/shared_common/shrinit.cpp]: I fixed it.

**Comment by keithc-ca** [comment]
Jenkins test sanity.functional  amac,win jdk21

**Comment by theresa-m** [comment]
Except for maybe the timeouts in SCURLClassLoader* tests these failures appear to be known issues https://github.com/eclipse-openj9/openj9/issues/19678  https://github.com/eclipse-openj9/openj9/issues/19408.

I'm rerunning windows sanity tests to check again https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/24027/

**Comment by pshipton** [comment]
Internal build (wa) passed http://vmfarm.rtp.raleigh.ibm.com/build_info.php?build_id=78507

File: runtime/shared_common/shrinit.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/shared_common/shrinit.cpp]: I don't remember seeing multiple threads calling into here before. A simple printf should be sufficient to prove there is only 1 thread here.

[the remark this answers is not in the document]
Comment (keithc-ca) [reply runtime/shared_common/shrinit.cpp]: There is only one call to `j9shr_shutdown()` which occurs as the VM is shutting down (or failed to start) and after other threads have been terminated.

**Comment by keithc-ca** [comment]
Jenkins test sanity.functional amac,win jdk21

**Pull request merged**